torch/distributed/pipelining/schedules.py

```diff
@@ -955,3 +956,151 @@ def backward_stage_index(step):
         for _ in range(self.pp_group_size - rank - 1):
             rank_ops.append(None)
         return rank_ops
+
+
+class ScheduleFlexibleInterleaved1F1B(PipelineScheduleMulti):
+    """
+    The Flexible Interleaved 1F1B schedule.
+    Supports some num_microbatch configurations where num_microbatch % pp_size != 0
+    Using the flex_pp schedule, we will have num_rounds = max(1, n_microbatches // pp_group_size) and
+    it works as long as n_microbatches % num_rounds is 0. As a few examples, support
+
+    1. pp_group_size = 4, n_microbatches = 10. We will have num_rounds = 2 and n_microbatches % 2 is 0.
+    2. pp_group_size = 4, n_microbatches = 3. We will have num_rounds = 1 and n_microbatches % 1 is 0.
```

Comment:
can you explain more what 'round' means in the doc? 

is it wasteful to have an extra round?

if user was on the fence between using pp_group_size=4 and n_microbatch = 3 vs 4, what are the tradeoffs they should consider?

Reply:
Great question! Let me try my best to explain this :)

Logically, I think you are grouping your microbatches into groups of k where k = pp_group_size to calculate forward_stage_index and backward_stage_index. However, k doesn't really have to be the same as pp_group_size. Suppose pp_group_size = 4 and n_microbatches = 10, we can have k = 2 and group the batches into 2 logical groups each with 5 items instead.